radar/src/radar.rs

```diff
@@ -0,0 +1,603 @@
+                    let payload = serde_json::json!({ "params": rc });
+                    publish_message(ctx, set_subject, payload)?;
+                }
+            }
+        }
+    }
+    Ok(vec![])
+}
+
+/// Helper function used to publish a payload on a specified subjct
+fn publish_message(
+    ctx: &CapabilitiesContext,
+    subject: &str,
+    payload: serde_json::value::Value,
+) -> CallResult {
+    if ctx
+        .msg()
+        .publish(subject, None, &serde_json::to_vec(&payload)?)
+        .is_err()
+    {
+        Err("Error publishing message".into())
+    } else {
+        Ok(vec![])
+    }
+}
+
+/// Function to compute all changes to a contact list needed given a resources id, current position,
+/// radar receiver, all old contacts, and a map of all entity positions that are published.
+/// Changes are in the form of RadarContactDeltas, either specifying to Add, Remove, or Change a contact.
+fn radar_updates(
+    entity_id: &str,
+    current_position: &Position,
+    radar_receiver: &RadarReceiver,
+    old_contacts: &HashMap<String, RadarContact>,
+    all_positions: &HashMap<String, Position>,
+) -> Vec<RadarContactDelta> {
+    let mut deltas: Vec<RadarContactDelta> = Vec::new();
+    let contacts: Vec<String> = old_contacts
+        .values()
+        .map(|rc| rc.clone().entity_id)
+        .collect();
+    for (k, v) in all_positions.iter() {
+        if contacts.contains(k) {
+            let mut rid: String = "".to_string();
+            for (key, val) in old_contacts.iter() {
+                if val.entity_id == *k {
+                    rid = key.to_string();
+                    break;
+                }
+            }
+            if within_radius(current_position, v, radar_receiver.radius) {
+                let vector_to = current_position.vector_to(v);
+                deltas.push(RadarContactDelta::Change(
+                    rid.replace(":", "."),
+                    RadarContact {
+                        entity_id: k.clone().to_string(),
+                        distance: vector_to.mag,
+                        heading_xy: vector_to.heading_xy,
+                        heading_z: vector_to.heading_z,
+                    },
+                ));
+            } else {
+                deltas.push(RadarContactDelta::Remove(rid));
+            }
+        } else if entity_id != k && within_radius(current_position, &v, radar_receiver.radius) {
+            let vector_to = current_position.vector_to(v);
+            deltas.push(RadarContactDelta::Add(RadarContact {
+                entity_id: k.clone().to_string(),
+                distance: vector_to.mag,
+                heading_xy: vector_to.heading_xy,
+                heading_z: vector_to.heading_z,
+            }));
+        }
+    }
+    deltas
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+enum RadarContactDelta {
+    Add(RadarContact),
+    Remove(String),
+    Change(String, RadarContact),
+}
```

Comment:
Does this belong here or in stacktrader components?

Reply:
This probably should be internal, since it's only produced and consumed by the radar system - no other system is exposed to this data structure.